### PR TITLE
[SPARK-18220][SQL] read Hive orc table with varchar column should not fail

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Spark SQL only has `StringType`, when reading hive table with varchar column, we will read that column as `StringType`. However, we still need to use varchar `ObjectInspector` to read varchar column in hive table, which means we need to know the actual column type at hive side.

In Spark 2.1, after https://github.com/apache/spark/pull/14363 , we parse hive type string to catalyst type, which means the actual column type at hive side is erased. Then we may use string `ObjectInspector` to read varchar column and fail.

This PR keeps the original hive column type string in the metadata of `StructField`, and use it when we convert it to a hive column.

## How was this patch tested?

newly added regression test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @gatorsmile

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala [hunk 1]
```diff
@@ -51,9 +51,12 @@ private[spark] object HiveUtils extends Logging {
     sc
   }
 
-  /** The version of hive used internally by Spark SQL. */
+  // The version of hive used internally by Spark SQL.
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: why did you change the comment style?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala [hunk 2]
```diff
@@ -51,9 +51,12 @@ private[spark] object HiveUtils extends Logging {
...
   val hiveExecutionVersion: String = "1.2.1"
 
+  // The property key that is used to store the raw hive type string in the metadata of StructField.
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: I'd add a bit more color here, e.g. by adding an example: "For example, in the case where the Hive type is varchar, the type gets mapped to a string type in Spark SQL, but we need to preserve the original type in order to invoke the correct object inspector in Hive"

**Comment by rxin** [comment]
Merging in master/branch-2.1.
